File: lib/engine/peerComparison.ts
import type { AnomalyResult, Direction } from '../types';

/** 同儕比較偏差門檻 */
const DEVIATION_THRESHOLD = 0.10;  // 10%
const MONITOR_THRESHOLD = 0.20;    // 20%（monitor 型指標）

/**
 * 偵測與同儕值的偏差
 * 差異率 = (本院值 - 同儕值) / 同儕值 × 100%
 */
export function detectPeerDeviation(
  value: number,
  peerValue: number | null,
  direction: Direction,
  year?: number,
  month?: number
): AnomalyResult | null {
  if (peerValue === null || peerValue === undefined || peerValue === 0) return null;

  const deviationRate = (value - peerValue) / peerValue;
  const absDeviation = Math.abs(deviationRate);
  const deviationPercent = (deviationRate * 100).toFixed(1);

  if (direction === 'lower') {
    // 越低越好
    if (value > peerValue * (1 + DEVIATION_THRESHOLD)) {
      // 高於同儕 — 不利
      return {
        mechanism: 'peer_comparison',
        severity: 'watch',
        direction: 'unfavorable',
        message: `高於同儕值 ${deviationPercent}%（同儕值: ${peerValue.toFixed(2)}）`,
        value,
        referenceValue: peerValue,
        year,
        month,
      };
    }
    if (value <= peerValue * (1 - DEVIATION_THRESHOLD)) {
      // 低於同儕 — 優良
      return {
        mechanism: 'peer_comparison',
        severity: 'excellent',
        direction: 'favorable',
        message: `低於同儕值 ${Math.abs(parseFloat(deviationPercent))}%（同儕值: ${peerValue.toFixed(2)}）`,
        value,
        referenceValue: peerValue,
        year,
        month,
      };
    }
  } else if (direction === 'higher') {
    // 越高越好
    if (value < peerValue * (1 - DEVIATION_THRESHOLD)) {
      // 低於同儕 — 不利
      return {
        mechanism: 'peer_comparison',
        severity: 'watch',
        direction: 'unfavorable',
        message: `低於同儕值 ${Math.abs(parseFloat(deviationPercent))}%（同儕值: ${peerValue.toFixed(2)}）`,
        value,
        referenceValue: peerValue,
        year,
        month,
      };
    }
    if (value >= peerValue * (1 + DEVIATION_THRESHOLD)) {
      // 高於同儕 — 優良
      return {
        mechanism: 'peer_comparison',
        severity: 'excellent',
        direction: 'favorable',
        message: `高於同儕值 ${deviationPercent}%（同儕值: ${peerValue.toFixed(2)}）`,
        value,
        referenceValue: peerValue,
        year,
        month,
      };
    }
  } else {
    // monitor：差異率絕對值 > 20% 為異常
    if (absDeviation > MONITOR_THRESHOLD) {
      return {
        mechanism: 'peer_comparison',
        severity: 'watch',
        direction: 'unfavorable',
        message: `與同儕值差異 ${deviationPercent}%（同儕值: ${peerValue.toFixed(2)}）`,
        value,
        referenceValue: peerValue,
        year,
        month,
      };
    }
  }

  return null;
}

File: backend/apps/analysis/services/peer_comparison.py
"""同儕值比較偵測"""
from __future__ import annotations

from .control_chart import AnomalyResult

DEVIATION_THRESHOLD = 0.10  # 10%
MONITOR_THRESHOLD = 0.20    # 20%


def detect_peer_deviation(
    value: float,
    peer_value: float | None,
    direction: str,
    year: int | None = None,
    month: int | None = None,
) -> AnomalyResult | None:
    """偵測與同儕值的偏差"""
    if peer_value is None or peer_value == 0:
        return None

    deviation_rate = (value - peer_value) / peer_value
    abs_deviation = abs(deviation_rate)
    dev_pct = f"{deviation_rate * 100:.1f}"

    if direction == "lower":
        if value > peer_value * (1 + DEVIATION_THRESHOLD):
            return AnomalyResult(
                mechanism="peer_comparison", severity="watch", direction="unfavorable",
                message=f"高於同儕值 {dev_pct}%（同儕值: {peer_value:.2f}）",
                value=value, reference_value=peer_value, year=year, month=month,
            )
        if value <= peer_value * (1 - DEVIATION_THRESHOLD):
            return AnomalyResult(
                mechanism="peer_comparison", severity="excellent", direction="favorable",
                message=f"低於同儕值 {abs(float(dev_pct))}%（同儕值: {peer_value:.2f}）",
                value=value, reference_value=peer_value, year=year, month=month,
            )
    elif direction == "higher":
        if value < peer_value * (1 - DEVIATION_THRESHOLD):
            return AnomalyResult(
                mechanism="peer_comparison", severity="watch", direction="unfavorable",
                message=f"低於同儕值 {abs(float(dev_pct))}%（同儕值: {peer_value:.2f}）",
                value=value, reference_value=peer_value, year=year, month=month,
            )
        if value >= peer_value * (1 + DEVIATION_THRESHOLD):
            return AnomalyResult(
                mechanism="peer_comparison", severity="excellent", direction="favorable",
                message=f"高於同儕值 {dev_pct}%（同儕值: {peer_value:.2f}）",
                value=value, reference_value=peer_value, year=year, month=month,
            )
    else:  # monitor
        if abs_deviation > MONITOR_THRESHOLD:
            return AnomalyResult(
                mechanism="peer_comparison", severity="watch", direction="unfavorable",
                message=f"與同儕值差異 {dev_pct}%（同儕值: {peer_value:.2f}）",
                value=value, reference_value=peer_value, year=year, month=month,
            )

    return None
